From 2755bb84f22adc4ea986cc75187e52d7eac618f4 Mon Sep 17 00:00:00 2001
From: gstavi <gur.stavi@gmail.com>
Date: Wed, 15 Jul 2015 16:41:20 +0300
Subject: [PATCH 08/17] prevent static windows to be drawn outside of monitor.
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="------------2.30.0"

This is a multi-part message in MIME format.
--------------2.30.0
Content-Type: text/plain; charset=UTF-8; format=fixed
Content-Transfer-Encoding: 8bit


* Mainly for file list on ctrl-tab when many files are open.

diff --git a/PowerEditor/src/WinControls/StaticDialog/StaticDialog.cpp b/PowerEditor/src/WinControls/StaticDialog/StaticDialog.cpp
index 425b1885..41ef8839 100644
--- a/PowerEditor/src/WinControls/StaticDialog/StaticDialog.cpp
+++ b/PowerEditor/src/WinControls/StaticDialog/StaticDialog.cpp
@@ -62,19 +62,72 @@ POINT StaticDialog::getTopPoint(HWND hwnd, bool isLeft) const
 	return p;
 }
 
+static int get_monitor_rect(HWND hwnd, RECT *rect)
+{
+	HMONITOR MonitorHndl;
+	MONITORINFO MonitorInfo;
+	BOOL res;
+
+	MonitorHndl = ::MonitorFromWindow (hwnd, MONITOR_DEFAULTTONEAREST);
+	::memset(&MonitorInfo, 0, sizeof(MonitorInfo));
+	MonitorInfo.cbSize = sizeof(MonitorInfo);
+	res = ::GetMonitorInfo(MonitorHndl, &MonitorInfo);
+	if (res == 0)
+		return 0;
+	*rect = MonitorInfo.rcWork;
+	return 1;
+}
+
+/*
+ * Gur: prevent static windows to be drawn outside of monitor. Mainly for file
+ * list on ctrl-tab when many files are open.
+ */
+
 void StaticDialog::goToCenter()
 {
 	RECT rc;
+	int x, y, width, height;
 	::GetClientRect(_hParent, &rc);
 	POINT center;
 	center.x = rc.left + (rc.right - rc.left)/2;
 	center.y = rc.top + (rc.bottom - rc.top)/2;
 	::ClientToScreen(_hParent, &center);
 
-	int x = center.x - (_rc.right - _rc.left)/2;
-	int y = center.y - (_rc.bottom - _rc.top)/2;
-
-	::SetWindowPos(_hSelf, HWND_TOP, x, y, _rc.right - _rc.left, _rc.bottom - _rc.top, SWP_SHOWWINDOW);
+	width = _rc.right - _rc.left;
+	height = _rc.bottom - _rc.top;
+
+	x = center.x - width/2;
+	y = center.y - height/2;
+
+	if (y < 0)
+		y = 0;
+	if (x < 0)
+		x = 0;
+	if (get_monitor_rect(_hParent, &rc))
+	{
+		int mon_width, mon_height;
+		mon_width = rc.right - rc.left;
+		mon_height = rc.bottom - rc.top;
+		if (width > mon_width)
+		{
+			x = 0;
+			width = mon_width;
+		}
+		else if (x + width > mon_width)
+		{
+			x = mon_width - width;
+		}
+		if (height > mon_height)
+		{
+			y = 0;
+			height = mon_height;
+		}
+		else if (y + height > mon_height)
+		{
+			y = mon_height - height;
+		}
+	}
+	::SetWindowPos(_hSelf, HWND_TOP, x, y, width, height, SWP_SHOWWINDOW);
 }
 
 void StaticDialog::display(bool toShow, bool enhancedPositioningCheckWhenShowing) const

--------------2.30.0--


